File: apps/stream/app/schemas/stream.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator, model_validator

from app.models.stream import ChannelType, StreamMode, StreamStatus

# --- Dispatch schemas ---

# --- Channel config models (for API validation) ---


class TeamsConfig(BaseModel):
    workflow_url: HttpUrl = Field(..., description="Microsoft Teams Power Automate workflow URL")


class SlackWebhookConfig(BaseModel):
    webhook_url: HttpUrl = Field(..., description="Slack incoming webhook URL")


class WebhookConfig(BaseModel):
    url: HttpUrl = Field(..., description="Target webhook URL")
    method: str = Field("POST", description="HTTP method (POST or PUT)")
    headers: dict[str, str] = Field(default_factory=dict, description="Custom headers to include")
    secret: str | None = Field(None, description="Shared secret for HMAC signature verification")


CHANNEL_CONFIG_MAP: dict[ChannelType, type[BaseModel]] = {
    ChannelType.TEAMS: TeamsConfig,
    ChannelType.SLACK_WEBHOOK: SlackWebhookConfig,
    ChannelType.WEBHOOK: WebhookConfig,
}

# --- Stream schemas ---


class StreamCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255, description="Stream name")
    description: str | None = Field(None, description="Optional description")
    channel_type: ChannelType
    channel_config: dict[str, Any] = Field(..., description="Channel-specific configuration")
    mode: StreamMode
    cron_expression: str | None = Field(None, description="Cron expression (required for recurrence mode)")
    subscribed_events: list[str] = Field(default_factory=list, description="Event types to subscribe to")

    @field_validator("subscribed_events", mode="before")
    @classmethod
    def validate_subscribed_events(cls, v: list) -> list:
        for item in v:
            parts = str(item).split(".")
            if len(parts) != 3 or not all(p.strip() for p in parts):
                raise ValueError(f"subscribed_events item '{item}' must follow source.resource.action format")
        return v

    @model_validator(mode="after")
    def validate_mode_and_channel(self) -> "StreamCreate":
        if self.mode == StreamMode.RECURRENCE and not self.cron_expression:
            raise ValueError("cron_expression is required for recurrence mode")
        if self.mode == StreamMode.LIVE and self.cron_expression:
            raise ValueError("cron_expression is forbidden for live mode")

        config_validators = {
            ChannelType.TEAMS: TeamsConfig,
            ChannelType.SLACK_WEBHOOK: SlackWebhookConfig,
            ChannelType.WEBHOOK: WebhookConfig,
        }
        validator = config_validators.get(self.channel_type)
        if validator:
            try:
                validator(**self.channel_config)
            except Exception as exc:
                raise ValueError(f"channel_config is invalid for {self.channel_type}: {exc}") from exc
        return self


class StreamUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    description: str | None = None
    channel_config: dict[str, Any] | None = None
    cron_expression: str | None = None
    subscribed_events: list[str] | None = None


class StreamStatusUpdate(BaseModel):
    status: StreamStatus = Field(..., description="New status for the stream")


class StreamRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    folder_id: str
    channel_type: ChannelType
    channel_config: dict[str, Any]
    mode: StreamMode
    cron_expression: str | None
    status: StreamStatus
    organization_id: str
    owner_id: str
    owner_username: str | None
    subscribed_events: list[str]
    created_at: datetime
    updated_at: datetime


# --- Test connection / dispatch schemas ---


class TestConnectionRequest(BaseModel):
    channel_type: ChannelType
    channel_config: dict[str, Any] = Field(..., description="Channel-specific configuration to test")


class TestConnectionResponse(BaseModel):
    success: bool
    error: str | None = None


class DispatchResponse(BaseModel):
    dispatched_count: int
    failed_count: int

File: apps/front/src/types/stream.ts
/**
 * Stream types for multi-channel event distribution.
 *
 * Streams deliver folder events (company created, task completed, etc.)
 * to external channels like Teams, Slack, or generic webhooks.
 */

// Channel types — matches backend ChannelType enum
export type ChannelType = 'teams' | 'slack_webhook' | 'webhook'

// Stream execution modes
export type StreamMode = 'live' | 'recurrence'

// Stream lifecycle statuses
export type StreamStatus = 'draft' | 'active' | 'paused' | 'archived'

// Delivery statuses for individual event deliveries
export type DeliveryStatus = 'pending' | 'delivered' | 'failed' | 'skipped'

// --- Channel config shapes ---

/**
 * Microsoft Teams Power Automate workflow configuration.
 */
export interface TeamsConfig {
  workflow_url: string
}

/**
 * Slack incoming webhook configuration.
 */
export interface SlackWebhookConfig {
  webhook_url: string
}

/**
 * Generic HTTP webhook configuration.
 */
export interface WebhookConfig {
  url: string
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE'
  headers?: Record<string, string>
  secret?: string | null
}

/**
 * Maps each channel type to its configuration shape.
 *
 * Allows type-safe constraints like `ChannelConfigMap['teams']` → `TeamsConfig`.
 * Used by `StreamCreate` (via `StreamChannel`) and `StreamUpdate<T>` to ensure
 * `channel_config` matches the actual `channel_type`.
 */
export interface ChannelConfigMap {
  teams: TeamsConfig
  slack_webhook: SlackWebhookConfig
  webhook: WebhookConfig
}

// Union of all channel configs (kept as a convenience alias)
export type ChannelConfig = ChannelConfigMap[ChannelType]

/**
 * Discriminated pair of {channel_type, channel_config}: picking a channel_type
 * automatically narrows channel_config to its matching shape.
 */
export type StreamChannel = {
  [K in ChannelType]: { channel_type: K; channel_config: ChannelConfigMap[K] }
}[ChannelType]

/**
 * Discriminated schedule: cron_expression is required in recurrence mode and
 * forbidden in live mode.
 */
export type StreamSchedule =
  | { mode: 'live'; cron_expression?: null }
  | { mode: 'recurrence'; cron_expression: string }

// --- API request/response shapes ---

/**
 * Full stream response from the API.
 */
export interface StreamRead {
  id: number
  name: string
  description: string | null
  folder_id: string
  channel_type: ChannelType
  channel_config: ChannelConfig
  mode: StreamMode
  cron_expression: string | null
  status: StreamStatus
  organization_id: string
  owner_id: string
  owner_username: string | null
  subscribed_events: string[]
  created_at: string
  updated_at: string
}

/**
 * Request body for creating a stream.
 *
 * Discriminated on two axes:
 *   - `channel_type` constrains `channel_config` to its matching shape
 *   - `mode` decides whether `cron_expression` is required
 */
export type StreamCreate = {
  name: string
  description?: string | null
  subscribed_events?: string[]
} & StreamChannel &
  StreamSchedule

/**
 * Request body for updating a stream (all fields optional).
 * Note: channel_type and mode are NOT updatable.
 *
 * Generic parameter `T` lets callers constrain `channel_config` to the stream's
 * actual channel type: `StreamUpdate<'teams'>` forbids sending a SlackConfig.
 * Defaults to the full union for callers that don't know the type statically.
 */
export interface StreamUpdate<T extends ChannelType = ChannelType> {
  name?: string | null
  description?: string | null
  channel_config?: ChannelConfigMap[T] | null
  cron_expression?: string | null
  subscribed_events?: string[] | null
}

/**
 * Request body for updating stream status.
 */
export interface StreamStatusUpdate {
  status: StreamStatus
}

/**
 * Individual delivery record for a stream event.
 */
export interface DeliveryRead {
  id: number
  stream_id: number
  event_id: number
  status: DeliveryStatus
  error_message: string | null
  response_metadata: Record<string, unknown> | null
  delivered_at: string | null
  created_at: string
  attempt_count: number
  next_retry_at: string | null
}

// --- Event types catalog ---

/**
 * Single event type entry.
 */
export interface EventTypeEntry {
  type: string
  label: string
}

/**
 * Group of event types from a source module.
 */
export interface EventSourceGroup {
  source: string
  label: string
  available: boolean
  events: EventTypeEntry[]
}

// --- Test connection ---

/**
 * Request to test a channel connection.
 */
export interface TestConnectionRequest {
  channel_type: ChannelType
  channel_config: ChannelConfig
}

/**
 * Response from testing a channel connection.
 */
export interface TestConnectionResponse {
  success: boolean
  error?: string | null
}

// --- Dispatch ---

/**
 * Response from manual dispatch.
 */
export interface DispatchResponse {
  dispatched_count: number
  failed_count: number
}

// --- Builders ---

/**
 * Narrow a flat form payload into a type-safe {@link StreamCreate}.
 *
 * The form state carries a union `channel_type: ChannelType` which TypeScript
 * can't assign to a discriminated `StreamChannel`. This helper performs the
 * runtime narrowing once, at the form/API boundary.
 */
export const buildStreamCreate = (input: {
  name: string
  description: string | null
  channel_type: ChannelType
  channel_config: ChannelConfig
  mode: StreamMode
  cron_expression?: string | null
  subscribed_events: string[]
}): StreamCreate => {
  const base = {
    name: input.name,
    description: input.description,
    subscribed_events: input.subscribed_events,
  }

  const channel: StreamChannel =
    input.channel_type === 'teams'
      ? { channel_type: 'teams', channel_config: input.channel_config as TeamsConfig }
      : input.channel_type === 'slack_webhook'
        ? {
            channel_type: 'slack_webhook',
            channel_config: input.channel_config as SlackWebhookConfig,
          }
        : { channel_type: 'webhook', channel_config: input.channel_config as WebhookConfig }

  const schedule: StreamSchedule =
    input.mode === 'recurrence'
      ? { mode: 'recurrence', cron_expression: input.cron_expression ?? '' }
      : { mode: 'live' }

  return { ...base, ...channel, ...schedule }
}
